### PR TITLE
Update to branch before adding detection module

### PR DESCRIPTION
**Pull request by ricolandman**
Finally found the time to add all changes properly. What I added:
- obs_nodding_irregular: Function that works for any nodding sequence (I needed this for my own transit observations). I wanted to implement this within obs_nodding but can't think of a simple way to do it. Right now it is a lot of duplicate code, so could definitely be improved in the future.
- util_extract_science: Function that extracts the results from obs_nodding for specific slit_fraction (e.g. for widely separated planets).
- 2nd order wavelength correction. This "grid search" is quite memory intensive, so I run "correct_wavelengths" multiple times with decreasing accuracy to make it feasible.
- Changed initial guess in fit_gaussian to make the function also works when the star is not in the center.
- custom_extract_2d: Including correction for slit tilt. This makes sure the wavelength solution is consistent vertically, so that we can sum over the spatial dimension in correct_wavelengths_2d.
- obs_nodding: Added the option "check_existing" if you want to skip already reduced files.

I tested it on the beta Pic data and some transit observations and it all runs smooth, but can't guarantee that there are no bugs. Feel free to change stuff or let me know if you want things changed. I have started on the detection module, which should come soon.

### CODE REVIEW

**Comment by ricolandman** [comment]
Oops, I see I messed up the branches. This actually already includes the functions for planet detection:
- remove_starlight: Function for removing the stellar contribution from each row
- remove_pca: Function for removing PCA components for each row
- cross_correlate: Function for cross-correlation.

While these functions are definitely not optimized (I can get a much higher S/N detection with my custom method) you should be able to detect e.g. beta Pic b quite easily using these functions.

**Review by tomasstolker — CHANGES_REQUESTED** [review]
Amazing work!! Many thanks @ricolandman!

I tested the new methods and it runs all fine. I did not check all the code in detail but from what I can tell it all looks good! Just added a few minor comments. The only larger request that I would have is to add some brief docstrings to the `util` functions if you have the time for that.

Does the new `obs_nodding_irregular` method also work on a regular/ABBA nodding pattern? If so then feel free to remove the old `obs_nodding` and use `obs_nodding_irregular` as the new `obs_nodding` 👍.

Let me know in case you have any questions!

**Comment by tomasstolker** [comment]
Thanks a lot for the additional improvements!

The checks are not passing. I think because the `requirements.txt` may not be complete?

File: pycrires/util.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+import numpy as np
+from scipy.signal import savgol_filter
+from scipy.interpolate import interp1d
+import os
+from PyAstronomy.pyasl import fastRotBroad
+
+def lowpass_filter(flux, window_length):
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/util.py:8]: Done

File: pycrires/pipeline.py [hunk 2]
```diff
@@ -4587,99 +5032,285 @@ def correct_wavelengths_2d(
...
+
+            # Save the correlation plots
+            if create_plots:
+                fig.add_subplot(111, frame_on=False)
+                plt.tick_params(labelcolor="none", bottom=False, left=False)
+                plt.ylabel("Slope", fontsize=16)
+                plt.xlabel("Offset (nm)", fontsize=16)
+                out_label = fits_file[-17:-5]
+                plt.tight_layout()
+                plt.savefig(f"{output_dir}/correlation_map_{nod_ab}_{accuracy}.png")
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py [hunk 3]
```diff
@@ -5312,6 +5962,388 @@ def gaussian(x, amp, mean, sigma):
...
+
+        for item in out_files:
+            self._update_files(f"STAR_REMOVED_{nod_ab}", str(item))
+
+        with open(self.json_file, "w", encoding="utf-8") as json_file:
+            json.dump(self.file_dict, json_file, indent=4)
+
+
+    @typechecked
+    def remove_pca(
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py [hunk 4]
```diff
@@ -4486,22 +4925,22 @@ def correct_wavelengths(
                             plt.sca(axes[order, i_det])
                             plt.axis("off")
 
+                    print(opt_b, opt_a, opt_c)
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Yes

File: pycrires/pipeline.py [hunk 5]
```diff
@@ -5312,6 +5962,388 @@ def gaussian(x, amp, mean, sigma):
...
+            hdu_list = fits.open(fits_file)
+            spec = hdu_list['SPEC'].data
+            wl = hdu_list['WAVE'].data
+            err = hdu_list['ERR'].data
+            star_subtracted = np.zeros_like(spec)
+            for det_idx in np.arange(3):
+                for order_idx, (order_spec, order_wl, order_err) in enumerate(
+                    zip(spec[det_idx], wl[det_idx], err[det_idx])):
+
+                    print(f"Removing starlight for detector {det_idx}"
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/util.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+def flag_outliers(order_spec, sigma=4):
+    z = (order_spec - np.nanmedian(order_spec, axis=1)[:, np.newaxis])\
+                /np.nanstd(order_spec, axis=1)[:, np.newaxis]
+    outliers = z>sigma
+    order_spec[outliers] = np.nan
+    return order_spec
+
+
+def load_BT_SETTL_template(temperature, log_g, vsini, wl_lims=[0.5, 3]):
+    import wget
```
[the remark this answers is not in the document]
Comment (tomasstolker) [reply pycrires/util.py:190]: Okay no worries, I can have a look separately from this PR.

File: pycrires/pipeline.py [hunk 7]
```diff
@@ -4517,19 +4956,22 @@ def correct_wavelengths(
...
         self,
         nod_ab: str = "A",
-        accuracy: float = 0.01,
+        accuracy: float = 0.002,
         window_length: int = 201,
+        N_grid: int = 51,
         minimum_strength: float = 0.005,
         sum_over_spatial_dim: bool = True,
-        input_folder="fit_gaussian",
+        sum_over_time: bool = True,
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py [hunk 8]
```diff
@@ -4587,99 +5032,285 @@ def correct_wavelengths_2d(
...
         # Read extracted spectra
+        if input_folder=='custom_extract_2d':
+            fits_files = list(self.file_dict[f'CUSTOM_EXTRACT_2D_{nod_ab}'].keys())
+        elif input_folder=='util_extract_2d':
+            fits_files = list(self.file_dict[f'UTIL_EXTRACT_2D_{nod_ab}'].keys())
+        elif input_folder=='fit_gaussian':
+            fits_files = list(self.file_dict[f'FIT_GAUSSIAN_2D_{nod_ab}'].keys())
+        elif input_folder=='correct_wavelengths_2d':
+            fits_files = list(self.file_dict[f'CORRECT_WAVELENGTHS_2D_{nod_ab}'].keys())
+        import glob
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py [hunk 9]
```diff
@@ -4587,99 +5032,285 @@ def correct_wavelengths_2d(
...
-                        )
-
-                        corrected_wavel[det_idx, order_idx, :] = wavel_2d
+                    # Check if there are enough telluric features in this
+                    # wavelength range
+                    wl_mask = (transm_spec[:, 0] > np.min(wavel_2d)) * \
+                              (transm_spec[:, 0] < np.max(wavel_2d))
+                    template_std = np.std(transm_spec[wl_mask,1])
+                    if not template_std > minimum_strength:
+                            print("WARNING: Not enough telluric features to correct wavelength"
```
[the remark this answers is not in the document]
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

**Review by tomasstolker — APPROVED** [review]
Thanks again for all your work @ricolandman! The PR seems ready to be merged 🎉

**Pull request merged**